File: frontend/src/types/exam.ts
export type SourceType = "global_success" | "common_knowledge" | "cambridge";
export type ExamStatus = "draft" | "reviewed" | "ready";
export type ExportMode = "plain" | "answer_key";
export type Difficulty = "nhan_biet" | "thong_hieu" | "van_dung" | "hon_hop";

export interface RefOut {
  id: string;
  code: string;
}

export interface ExerciseTypeRefOut {
  id: string;
  code: string;
  name: string;
  has_passage: boolean;
}

export interface QuestionOption {
  label: string;
  text: string;
  is_correct: boolean;
}

export interface QuestionOut {
  id: string;
  order_no: number;
  prompt_text: string;
  passage_text: string | null;
  options: QuestionOption[] | null;
  answer_text: string;
  explanation: string;
  target_knowledge: string;
  level: RefOut;
  source_ref: string;
  warnings: string[];
  is_approved: boolean;
  is_locked: boolean;
  part_id: string | null;
}

export interface BlockPartOut {
  id: string;
  order_no: number;
  title: string;
  instruction: string | null;
  question_count: number;
  prompt_override: string | null;
}

export interface BlockPartCreateRequest {
  title: string;
  instruction?: string | null;
  question_count: number;
  prompt_override?: string | null;
}

export type BlockPartUpdateRequest = Partial<BlockPartCreateRequest>;

export interface BlockOut {
  id: string;
  order_no: number;
  exercise_type: ExerciseTypeRefOut;
  title: string;
  instruction: string | null;
  question_count: number;
  points: string;
  difficulty: Difficulty;
  level_override: RefOut | null;
  shuffle_questions: boolean;
  shuffle_answers: boolean;
  prompt_override: string | null;
  passage_word_target: number | null;
  questions: QuestionOut[];
  parts: BlockPartOut[];
}

export interface ExamSummaryOut {
  id: string;
  title: string;
  status: ExamStatus;
  grade_number: number;
  level_code: string;
  total_questions: number;
  total_points: string;
  export_mode: ExportMode | null;
  variant_count: number;
  updated_at: string;
}

export interface ExamDetailOut {
  id: string;
  title: string;
  status: ExamStatus;
  source_type: SourceType;
  grade_id: string;
  level: RefOut;
  unit_id: string | null;
  grammar_topic_id: string | null;
  cambridge_certificate_id: string | null;
  extra_prompt: string | null;
  export_mode: ExportMode | null;
  variant_count: number;
  grammar_point_ids: string[];
  blocks: BlockOut[];
}

export interface ExamCreateRequest {
  title: string;
  grade_id: string;
  level_id: string;
  source_type: SourceType;
  unit_id?: string | null;
  grammar_topic_id?: string | null;
  cambridge_certificate_id?: string | null;
  extra_prompt?: string | null;
}

export interface BlockCreateRequest {
  exercise_type_id: string;
  title: string;
  instruction?: string | null;
  question_count: number;
  points: number;
  difficulty?: Difficulty;
  level_override_id?: string | null;
  shuffle_questions?: boolean;
  shuffle_answers?: boolean;
  prompt_override?: string | null;
  passage_word_target?: number | null;
}

export type BlockUpdateRequest = Partial<BlockCreateRequest>;

export interface QuestionFlagsUpdateRequest {
  is_approved?: boolean;
  is_locked?: boolean;
}

export interface ExportConfigRequest {
  export_mode: ExportMode;
  variant_count: number;
}

File: backend/app/schemas/exam.py
import uuid
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.exam import Difficulty, ExamStatus, ExportMode, SourceType


class ExamCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    grade_id: uuid.UUID
    level_id: uuid.UUID
    source_type: SourceType
    unit_id: uuid.UUID | None = None
    grammar_topic_id: uuid.UUID | None = None
    cambridge_certificate_id: uuid.UUID | None = None
    extra_prompt: str | None = None


class ExamUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    grade_id: uuid.UUID | None = None
    level_id: uuid.UUID | None = None
    source_type: SourceType | None = None
    unit_id: uuid.UUID | None = None
    grammar_topic_id: uuid.UUID | None = None
    cambridge_certificate_id: uuid.UUID | None = None
    extra_prompt: str | None = None


class GrammarSelectionRequest(BaseModel):
    grammar_point_ids: list[uuid.UUID]


class BlockCreateRequest(BaseModel):
    exercise_type_id: uuid.UUID
    title: str = Field(min_length=1, max_length=255)
    instruction: str | None = None
    question_count: int = Field(ge=1, le=50)
    points: Decimal = Field(ge=0, le=10)
    difficulty: Difficulty = Difficulty.HON_HOP
    level_override_id: uuid.UUID | None = None
    shuffle_questions: bool = True
    shuffle_answers: bool = True
    prompt_override: str | None = None
    passage_word_target: int | None = Field(default=None, ge=10, le=500)


class BlockUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    instruction: str | None = None
    question_count: int | None = Field(default=None, ge=1, le=50)
    points: Decimal | None = Field(default=None, ge=0, le=10)
    difficulty: Difficulty | None = None
    level_override_id: uuid.UUID | None = None
    shuffle_questions: bool | None = None
    shuffle_answers: bool | None = None
    prompt_override: str | None = None
    passage_word_target: int | None = Field(default=None, ge=10, le=500)


class BlockReorderRequest(BaseModel):
    block_ids: list[uuid.UUID]


class BlockPartCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    instruction: str | None = None
    question_count: int = Field(ge=1, le=50)
    prompt_override: str | None = None


class BlockPartUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    instruction: str | None = None
    question_count: int | None = Field(default=None, ge=1, le=50)
    prompt_override: str | None = None


class BlockPartReorderRequest(BaseModel):
    part_ids: list[uuid.UUID]


class QuestionFlagsUpdateRequest(BaseModel):
    is_approved: bool | None = None
    is_locked: bool | None = None


class ExportConfigRequest(BaseModel):
    export_mode: ExportMode
    variant_count: int = Field(ge=1, le=4)


class RefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    code: str


class QuestionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_no: int
    prompt_text: str
    passage_text: str | None
    options: list[dict[str, Any]] | None
    answer_text: str
    explanation: str
    target_knowledge: str
    level: RefOut
    source_ref: str
    warnings: list[str]
    is_approved: bool
    is_locked: bool
    part_id: uuid.UUID | None


class ExerciseTypeRefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    code: str
    name: str
    has_passage: bool


class BlockPartOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_no: int
    title: str
    instruction: str | None
    question_count: int
    prompt_override: str | None


class BlockOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_no: int
    exercise_type: ExerciseTypeRefOut
    title: str
    instruction: str | None
    question_count: int
    points: Decimal
    difficulty: Difficulty
    level_override: RefOut | None
    shuffle_questions: bool
    shuffle_answers: bool
    prompt_override: str | None
    passage_word_target: int | None
    questions: list[QuestionOut]
    parts: list[BlockPartOut]


class ExamSummaryOut(BaseModel):
    id: uuid.UUID
    title: str
    status: ExamStatus
    grade_number: int
    level_code: str
    total_questions: int
    total_points: Decimal
    export_mode: ExportMode | None
    variant_count: int
    updated_at: Any


class ExamDetailOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    status: ExamStatus
    source_type: SourceType
    grade_id: uuid.UUID
    level: RefOut
    unit_id: uuid.UUID | None
    grammar_topic_id: uuid.UUID | None
    cambridge_certificate_id: uuid.UUID | None
    extra_prompt: str | None
    export_mode: ExportMode | None
    variant_count: int
    grammar_point_ids: list[uuid.UUID]
    blocks: list[BlockOut]


class VariantOut(BaseModel):
    code: str
    seed: int
